### PR TITLE
Snippets

### PR DESCRIPTION
**Pull request by adamdriscoll**

### CODE REVIEW

**Comment by adamdriscoll** [comment]
Fixes #50

**Comment by adamdriscoll** [comment]
@AndreSayreMSFT can you review when you get a chance?

File: PowerShellTools/Project/ProvideCodeExpansionPathAttribute.cs [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+{
+
+    /// <summary>
+    /// This attribute registers an additional path for code snippets to live
+    /// in for a particular language.
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Class, AllowMultiple = true, Inherited = true)]
+    // Disable the "IdentifiersShouldNotHaveIncorrectSuffix" warning.
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Naming", "CA1711")]
+    sealed class ProvideCodeExpansionPathAttribute : RegistrationAttribute
```
Comment (AndreSayreMSFT) [inline PowerShellTools/Project/ProvideCodeExpansionPathAttribute.cs]: nit: explicit about public/internal/private acessor

Comment (AndreSayreMSFT) [inline PowerShellTools/Project/ProvideCodeExpansionPathAttribute.cs]: extended comment:This was existing code from PTVS, so my complaints lay with the original code :)

File: PowerShellTools/Project/ProvideCodeExpansionPathAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+//    implied. See the License for the specific language governing 
+//    permissions and limitations under the License.
+//
+//*********************************************************//
+
+using System;
+using System.Globalization;
+using Microsoft.VisualStudio.Shell;
+
+namespace Microsoft.VisualStudioTools
```
Comment (AndreSayreMSFT) [inline PowerShellTools/Project/ProvideCodeExpansionPathAttribute.cs]: Do we want to use a namespace for poshtools?

File: PowerShellTools/Project/ProvideCodeExpansionsAttribute.cs [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+//    implied. See the License for the specific language governing 
+//    permissions and limitations under the License.
+//
+//*********************************************************//
+
+using System;
+using System.Globalization;
+using Microsoft.VisualStudio.Shell;
+
+namespace Microsoft.VisualStudioTools {
```
Comment (AndreSayreMSFT) [inline PowerShellTools/Project/ProvideCodeExpansionsAttribute.cs]: ns?

File: PowerShellTools/Snippets/SnippetHandler.cs [hunk 4]
```diff
@@ -0,0 +1,576 @@
...
+
+namespace PowerShellTools.Snippets
+{
+    [Export(typeof(IVsTextViewCreationListener))]
+    [Name("snippets")]
+    [ContentType("PowerShell")]
+    [TextViewRole(PredefinedTextViewRoles.Editable)]
+    public class SnippetHandlerProvider : IVsTextViewCreationListener
+    {
+       // internal const string LanguageServiceGuidStr = "AD4D401C-11EA-431F-A412-FAB167156206";
```
Comment (AndreSayreMSFT) [inline PowerShellTools/Snippets/SnippetHandler.cs]: ?

File: PowerShellTools/Snippets/SnippetHandler.cs [hunk 5]
```diff
@@ -0,0 +1,576 @@
...
+                            // we want to leave the pass statement selected so the user can just
+                            // continue typing over it...
+                            var startLine = _textView.TextBuffer.CurrentSnapshot.GetLineFromPosition(startPosition + selectedIndex);
+                            _selectEndSpan = true;
+                            endSpan = new TextSpan()
+                            {
+                                iStartLine = startLine.LineNumber,
+                                iEndLine = startLine.LineNumber,
+                                iStartIndex = baseIndentation.Length + indentation.Length,
+                                iEndIndex = baseIndentation.Length + indentation.Length + 4,
```
Comment (AndreSayreMSFT) [inline PowerShellTools/Snippets/SnippetHandler.cs]: Q: This '4', will we run into nay funkiness when a user has tabs vs spaces?

File: PowerShellTools/Snippets/SnippetHandler.cs [hunk 6]
```diff
@@ -0,0 +1,576 @@
...
+                }
+            }
+
+            // get the indentation of where we're inserting the code...
+            string baseIndentation = GetBaseIndentation(ts);
+
+            TextSpan? endSpan = null;
+            using (var edit = _textView.TextBuffer.CreateEdit())
+            {
+                if (surroundsWith || surroundsWithStatement)
```
Comment (AndreSayreMSFT) [inline PowerShellTools/Snippets/SnippetHandler.cs]: Python and comments?

**Comment by AndreSayreMSFT** [comment]
Few minor comments, but otherwise looks good :)

**Pull request merged**